File: apps/web/src/components/flow-canvas/utils/handleTypes.ts
/**
 * Client-side mirror of apps/agent-service/src/domain/flows/handle_types.py.
 *
 * This is not a Langflow port — Langflow has no equivalent concept (its
 * component fields carry a fixed, hardcoded set of allowed input types
 * rather than a shared 9-value PortType lattice). This module exists so
 * illegal connections are rejected in `onConnect`, client-side, with no
 * request issued (design spec §10) — the server (P1 Task 3) re-validates
 * regardless and is the actual authority.
 *
 * The order below is load-bearing and mirrors the backend exactly:
 * identity -> Trigger isolation -> Data catch-all -> Data-as-source
 * rejection -> explicit coercions -> illegal. Trigger isolation outranking
 * the Data catch-all was a real bug on the backend (design spec amendment
 * F) — a Trigger is a pure signal with no payload, so it must not satisfy
 * a Data port just because Data otherwise accepts anything.
 *
 * See handleTypes.test.ts's 22.6 for how this is kept from silently
 * drifting out of sync with the backend.
 *
 * Brief: .tmp/flow-canvas-task-22-brief.md
 */

import type { PortType } from "../types/flow";

const EXPLICIT_COMPATIBLE: ReadonlySet<string> = new Set([
  serializePair("Message", "Text"),
]);

function serializePair(source: PortType, target: PortType): string {
  return `${source}->${target}`;
}

export function isCompatible(source: PortType, target: PortType): boolean {
  if (source === target) return true;
  if (source === "Trigger" || target === "Trigger") return false;
  if (target === "Data") return true;
  if (source === "Data") return false;
  return EXPLICIT_COMPATIBLE.has(serializePair(source, target));
}

/** A handle carries `types: PortType[]` (P1's Handle model) — compatible
 * if any pair across the two lists is compatible. */
export function handlesCompatible(
  sourceTypes: PortType[],
  targetTypes: PortType[]
): boolean {
  return sourceTypes.some((s) => targetTypes.some((t) => isCompatible(s, t)));
}

File: apps/agent-service/src/domain/flows/handle_types.py
"""Handle type compatibility.

Pure and total: every pair of ``PortType`` values returns a boolean, never
raises. Coercions are deliberately asymmetric — see
``.tmp/flow-canvas-design.md`` section 4.7. Silent coercion is how visual
builders produce flows that validate and then behave unexpectedly at runtime.

Adding a new ``PortType`` needs no change here to stay total (identity and the
``Data`` rules cover it automatically); it only needs an entry in
``_EXPLICIT_COMPATIBLE`` if it should deliberately coerce into something else.
That is a product decision each time, not a default.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Final

from models.flows import PortType

_EXPLICIT_COMPATIBLE: Final[frozenset[tuple[PortType, PortType]]] = frozenset(
    {
        (PortType.MESSAGE, PortType.TEXT),
    }
)


def is_compatible(source: PortType, target: PortType) -> bool:
    """Whether a value of type ``source`` may flow into a ``target`` port.

    ``Trigger`` isolation takes priority over the ``Data`` catch-all: a
    Trigger is a pure signal with no payload, so it does not satisfy a Data
    port just because Data otherwise accepts anything.
    """
    if source == target:
        return True
    if source is PortType.TRIGGER or target is PortType.TRIGGER:
        return False
    if target is PortType.DATA:
        return True
    if source is PortType.DATA:
        return False
    return (source, target) in _EXPLICIT_COMPATIBLE


def handles_compatible(source_types: Sequence[PortType], target_types: Sequence[PortType]) -> bool:
    """Whether any type on the source handle may flow into any type on the target."""
    return any(is_compatible(s, t) for s in source_types for t in target_types)
